src/persist-client/src/internal/state.rs

```diff
@@ -176,12 +176,18 @@ pub struct HandleDebugState {
 #[serde(tag = "type")]
 pub enum BatchPart<T> {
     Hollow(HollowBatchPart<T>),
+    Inline {
+        updates: LazyInlineBatchPart,
+        key_lower: Vec<u8>,
```

Comment:
I think it would probably be fine to omit the `key_lower`, in case that seems appealing! For tiny writes it might take up as much space as the write itself, and the main motivation for it (avoiding unnecessary fetches) doesn't really apply here.

Reply:
Yeah, I'd love to remove this!

I remember adding it because something in iter.rs was failing an assert. That was early last week and a lot was in flux in the branch, so could be that I'm remembering incorrectly or that it's no longer an issue. I'll try it again and see what happens.

In the meantime, were you thinking that we hardcode `vec![]` as the key_lower for inline parts or that we'd decode it on demand and return the real key lower? The first seems tricky because of the case where we have a run where the last part is inline (and it needs to have a greater key_lower than the previous ones, correct? Maybe this is the assert I was hitting). The second is a bit unfortunate because _the bytes are in there_ but prost won't let us get them out without allocs for key_offsets, val_offsets, ts, diff in the decoded proto, and I wasn't sure how hot a path key_lower is.

Reply:
> In the meantime, were you thinking [...]

The former!

> The first seems tricky because of the case where we have a run where the last part is inline [...]

I had not been thinking about this case! It seems a bit pathological to have a run with N>0 100MiB parts and then 1 inline part. We could probably ban it, though I appreciate that might be more work than just threading the lower around.

> and it needs to have a greater key_lower than the previous ones, correct?

That is certainly the case now; I don't know if anything _relies_ on it, though it seems plausible that we're asserting on it somewhere. If so we could probably relax such an assert, assuming that seems like the easiest path...

Reply:
Hardcode `vec![]` and disallow inline in a run if an earlier part is not inline sounds reasonable to me. I'll try typing it up and let you know how it goes

Reply:
Pushed it up, let's see what CI says

Reply:
CI and nightlies seem happy with it! \o/

Reply:
That works! (And should be easy enough to do the other half of it later if we decide it's important.)